File: src/main/java/me/tomassetti/antlrplus/ParserFacade.java
package me.tomassetti.antlrplus;

import org.antlr.v4.runtime.*;

import java.io.*;
import java.nio.charset.StandardCharsets;

public abstract class ParserFacade<R extends ParserRuleContext, P extends Parser> {

    public ParserFacade() {
    }

    protected abstract Lexer getLexer(ANTLRInputStream antlrInputStream);

    protected abstract P getParser(TokenStream tokens);

    protected abstract R getRoot(P parser);

    public R parseString(String code) {
        InputStream inputStream = new ByteArrayInputStream(code.getBytes(StandardCharsets.UTF_8));
        return parseStream(inputStream);
    }

    public R parseFile(File file) throws FileNotFoundException {
        return parseStream(new FileInputStream(file));
    }

    public R parseStream(InputStream inputStream) {
        try {
            Lexer lexer = getLexer(new org.antlr.v4.runtime.ANTLRInputStream(inputStream));
            TokenStream tokens = new CommonTokenStream(lexer);
            return getRoot(getParser(tokens));
        } catch (IOException e) {
            throw new RuntimeException("That is unexpected", e);
        }
    }
}

File: src/main/kotlin/me/tomassetti/antlrplus/ParserFacade.kt
package me.tomassetti.antlrplus

import org.antlr.v4.runtime.*

import java.io.*
import java.nio.charset.StandardCharsets

abstract class ParserFacade<R : ParserRuleContext, P : Parser> {

    protected abstract fun getLexer(antlrInputStream: ANTLRInputStream): Lexer

    protected abstract fun getParser(tokens: TokenStream): P

    protected abstract fun getRoot(parser: P): R

    fun parseString(code: String): R {
        val inputStream = ByteArrayInputStream(code.toByteArray(StandardCharsets.UTF_8))
        return parseStream(inputStream)
    }

    @Throws(FileNotFoundException::class)
    fun parseFile(file: File): R {
        return parseStream(FileInputStream(file))
    }

    fun parseStream(inputStream: InputStream): R {
        try {
            val lexer = getLexer(ANTLRInputStream(inputStream))
            val tokens = CommonTokenStream(lexer)
            return getRoot(getParser(tokens))
        } catch (e: IOException) {
            throw RuntimeException("That is unexpected", e)
        }

    }
}
